elasticapm/transport/base.py

```diff
@@ -53,12 +53,13 @@ def __init__(
         self._max_flush_time = max_flush_time
         self._max_buffer_size = max_buffer_size
         self._queued_data = None
-        self._flush_lock = threading.Lock()
+        self._queue_lock = threading.Lock()
         self._last_flush = timeit.default_timer()
         self._flush_timer = None
 
     def queue(self, event_type, data, flush=False):
-        self._queue(self.queued_data, {event_type: data})
+        with self._queue_lock:
+            self.queued_data.write((self._json_serializer({event_type: data}) + "\n").encode("utf-8"))
         since_last_flush = timeit.default_timer() - self._last_flush
         queue_size = self.queued_data_size
```

Comment:
`self.queued_data_size` calls `self.queued_data` so calls to `self.queued_data_size` must also be protected by the lock, or you lock inside `self.queued_data`

Reply:
Great catch. I'll remove the call to `self.queued_data` in `queued_data_size` and instead check for existence of `self._queued_data` directly and return `0` if it doesn't exist. Then we don't need a lock as it is read only. Makes sense?

Reply:
You could do something like:
```
with self._queue_lock:
            self.queued_data.write((self._json_serializer({event_type: data}) + "\n").encode("utf-8"))
            queue_size = self.queued_size

```

Reply:
@roncohen @simitt like this af2ffd490bc

Reply:
yeah that would probably work. It seems like `queue_size` here could end up being the size of a queue that was just flushed and we would then flush again, but that might not be an issue

Reply:
agree, this shouldn't really matter